File: src/0207-Course-Schedule/0207.cpp
class Solution 
{
public:
    bool canFinish(int n, vector<vector<int>>& pre) 
    {
        vector<vector<int>> g(n, vector<int>());
        vector<int> d(n, 0);
        vector<int> vis(n, 0);
        for (auto& p : pre) 
        {
            g[p[1]].push_back(p[0]);
            d[p[0]]++;
        }
        
        queue<int> q;
        for (int i = 0; i < n; i++) 
        {
            if (d[i] == 0) 
            {
                q.push(i);
                vis[i] = 1;
            }
        }
        while (!q.empty()) 
        {
            int cur = q.front(); q.pop(); n--;
            for (auto i : g[cur])
            {
                if (!vis[i] and --d[i] == 0) 
                {
                    vis[i] = 1;
                    q.push(i);
                }
            } 
        }
        return n == 0;
    }
};

File: src/0207-Course-Schedule/0207.go
func canFinish(n int, pre [][]int) bool {
    g := make(map[int][]int)
    d := make([]int, n)
    q := make([]int, 0)
    vis := make([]int, n)
    for _, v := range pre {
        g[v[1]] = append(g[v[1]], v[0])
        d[v[0]]++
    }
    
    for i := 0; i < n; i++ {
        if d[i] == 0 {
            q = append(q, i)
            vis[i] = 1
        }
    }
    
    for len(q) > 0 {
        cur := q[0]
        q = q[1:]
        n--
        for _, i := range g[cur] {
            d[i]--
            if vis[i] == 0 && d[i] == 0 {
                q = append(q, i)
                vis[i] = 1
            }
        }
    }
    return n == 0
}